cuda/test/base/lin_op.cu

```diff
@@ -39,6 +39,110 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 namespace {
 
 
+class DummyLinOp : public gko::EnableLinOp<DummyLinOp>,
+                   public gko::EnableCreateMethod<DummyLinOp> {
+public:
+    DummyLinOp(std::shared_ptr<const gko::Executor> exec,
+               gko::dim<2> size = gko::dim<2>{})
+        : EnableLinOp<DummyLinOp>(exec, size)
+    {}
+
+    void access() const { last_access = this->get_executor(); }
+
+    mutable std::shared_ptr<const gko::Executor> last_access;
+    mutable std::shared_ptr<const gko::Executor> last_b_access;
+    mutable std::shared_ptr<const gko::Executor> last_x_access;
+    mutable std::shared_ptr<const gko::Executor> last_alpha_access;
+    mutable std::shared_ptr<const gko::Executor> last_beta_access;
+
+protected:
+    void apply_impl(const gko::LinOp *b, gko::LinOp *x) const override
+    {
+        this->access();
+        static_cast<const DummyLinOp *>(b)->access();
+        static_cast<const DummyLinOp *>(x)->access();
+        last_b_access = b->get_executor();
+        last_x_access = x->get_executor();
+    }
+
+    void apply_impl(const gko::LinOp *alpha, const gko::LinOp *b,
+                    const gko::LinOp *beta, gko::LinOp *x) const override
+    {
+        this->access();
+        static_cast<const DummyLinOp *>(alpha)->access();
+        static_cast<const DummyLinOp *>(b)->access();
+        static_cast<const DummyLinOp *>(beta)->access();
+        static_cast<const DummyLinOp *>(x)->access();
+        last_alpha_access = alpha->get_executor();
+        last_b_access = b->get_executor();
+        last_beta_access = beta->get_executor();
+        last_x_access = x->get_executor();
+    }
+};
+
+
+class EnableLinOp : public ::testing::Test {
+protected:
+    EnableLinOp()
+        : ref{gko::ReferenceExecutor::create()},
+          cuda{gko::CudaExecutor::create(0, ref)},
+          op{DummyLinOp::create(cuda, gko::dim<2>{3, 5})},
+          alpha{DummyLinOp::create(ref, gko::dim<2>{1})},
+          beta{DummyLinOp::create(ref, gko::dim<2>{1})},
+          b{DummyLinOp::create(ref, gko::dim<2>{5, 4})},
+          x{DummyLinOp::create(ref, gko::dim<2>{3, 4})}
+    {}
+
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<const gko::CudaExecutor> cuda;
+    std::unique_ptr<DummyLinOp> op;
+    std::unique_ptr<DummyLinOp> alpha;
+    std::unique_ptr<DummyLinOp> beta;
+    std::unique_ptr<DummyLinOp> b;
+    std::unique_ptr<DummyLinOp> x;
+};
+
+
+TEST_F(EnableLinOp, ApplyCopiesDataToCorrectExecutor)
+{
+    op->apply(gko::lend(b), gko::lend(x));
+
+    ASSERT_EQ(op->last_b_access, cuda);
+    ASSERT_EQ(op->last_x_access, cuda);
+}
+
+
+TEST_F(EnableLinOp, ApplyCopiesBackOnlyX)
+{
+    op->apply(gko::lend(b), gko::lend(x));
+
+    ASSERT_EQ(b->last_access, nullptr);
```

Comment:
That's the same as what was present before. It's nullptr because `b` was not copied back. The access was done on the clone in the `cuda` executor, but that information wasn't propagated back to the original object which is sitting in the reference executor (because there is no need, it's read only).

Reply:
Thanks for explanation